### PR TITLE
Minesweeper: Port to GML compilation

### PR DESCRIPTION
**Pull request by tetektoza**
This patch ports minesweeper to GML compilation, and introduces
a few changes made to associated files.
It contributes to https://github.com/SerenityOS/serenity/issues/20518

### CODE REVIEW

**Comment by gmta** [comment]
Hi @tetektoza, welcome!

Could you please split up your commits so they individually contribute to either LibGUI or Minesweeper? Check [CONTRIBUTING.md](https://github.com/SerenityOS/serenity/blob/master/CONTRIBUTING.md#code-submission-policy)'s bullet on atomic commits.

**Comment by tetektoza** [comment]
Hi @gmta,

> Could you please split up your commits

Sure thing, should I also split up changes made to GMLCompiler (main.cpp file) to a separate commit? So that would be 3 separate commits.

**Comment by gmta** [comment]
> should I also split up changes made to GMLCompiler (main.cpp file) to a separate commit? So that would be 3 separate commits.

Sounds sensible! Improvements to the GMLCompiler are separate from the Minesweeper changes imo.

**Comment by tetektoza** [comment]
@gmta I think it should be fine now, you can check it out. All 3 commits are needed for the port to work, since some of the parameters specified in the .gml file weren't able to be converted implicitly. I need to split up my 2nd PR the same way but that will probably happen tomorrow.

**Comment by gmta** [comment]
@tetektoza The order of your commits is not right; you should first commit the changes to LibGUI so the next commit can use that feature. As it stands now, compilation fails for the first commit.

**Comment by tetektoza** [comment]
@gmta should be okay now, sorry for the mistake, now I get how it works :)

File: Userland/Games/Minesweeper/CMakeLists.txt [hunk 1]
```diff
@@ -4,19 +4,16 @@ serenity_component(
...
 
-stringify_gml(MinesweeperCustomGameWindow.gml MinesweeperCustomGameWindowGML.h minesweeper_custom_game_window_gml)
-stringify_gml(MinesweeperWindow.gml MinesweeperWindowGML.h minesweeper_window_gml)
+compile_gml(MainWidget.gml MainWidget.cpp)
+compile_gml(CustomGameWidget.gml CustomGameWidgetGML.cpp)
 
 set(SOURCES
-    CustomGameDialog.cpp
     Field.cpp
+    MainWidget.cpp
```
Comment (timschumi) [inline Userland/Games/Minesweeper/CMakeLists.txt:12]: Same change here.

File: Userland/Games/Minesweeper/CustomGameDialog.cpp [hunk 2]
```diff
@@ -38,21 +51,22 @@ void CustomGameDialog::set_max_mines()
     m_mines_spinbox->set_max((m_rows_spinbox->value() * m_columns_spinbox->value()) - 9);
 }
 
-CustomGameDialog::CustomGameDialog(Window* parent_window)
+CustomGameDialog::CustomGameDialog(NonnullRefPtr<Minesweeper::CustomGameWidget>
+                                       custom_game_widget,
```
[the remark this answers is not in the document]
Comment (timschumi) [reply Userland/Games/Minesweeper/CustomGameDialog.cpp:55]: That said, `CustomGameDialog` should probably be moved into the `Minesweeper` namespace as well, which alleviates the issue with too long type names.

**Review by timschumi — CHANGES_REQUESTED** [review]
Looking good, just a few remaining style fixups.

File: Userland/Games/Minesweeper/CMakeLists.txt (hunk 1 above)
Comment (tetektoza) [reply Userland/Games/Minesweeper/CMakeLists.txt]: Done.

File: Userland/Games/Minesweeper/CMakeLists.txt [hunk 3]
```diff
@@ -4,19 +4,16 @@ serenity_component(
...
+    CustomGameDialog.cpp
+    CustomGameWidgetGML.cpp
     main.cpp
```
[the remark this answers is not in the document]
Comment (tetektoza) [reply Userland/Games/Minesweeper/CMakeLists.txt]: Done.

**Review by timschumi — APPROVED** [review]

**Pull request merged**